editoast/src/views/pagination.rs

```diff
@@ -12,11 +12,85 @@ use serde::Serialize;
 use thiserror::Error;
 use tracing::warn;
 use utoipa::IntoParams;
+use utoipa::ToSchema;
 
 use crate::error::Result;
 use crate::modelsv2::DbConnection;
+use crate::ListAndCount;
+use crate::Model;
+use crate::SelectionSettings;
 
-const DEFAULT_PAGE_SIZE: i64 = 25;
+editoast_common::schemas! {
+    PaginationStats,
+}
+
+const DEFAULT_PAGE_SIZE: u64 = 25;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, ToSchema)]
+pub struct PaginationStats {
+    /// The total number of items
+    #[schema(minimum = 0)]
+    pub count: u64,
+
+    /// The number of items per page
+    #[schema(minimum = 1)]
+    pub page_size: u64,
+
+    /// The total number of pages
+    #[schema(minimum = 0)]
+    pub page_count: u64,
+
+    /// The current page number
+    #[schema(minimum = 1)]
+    pub current: u64,
+
+    /// The previous page number, if any
+    #[schema(required, minimum = 1)]
+    pub previous: Option<u64>,
+
+    /// The next page number, if any
+    #[schema(required, minimum = 1)]
+    pub next: Option<u64>,
+}
+
+impl PaginationStats {
+    pub fn new(current_page_count: u64, total_count: u64, page: u64, page_size: u64) -> Self {
+        assert!(page > 0);
+        assert!(page_size > 0);
+        assert!((page - 1) * page_size + current_page_count <= total_count);
```

Comment:
How about making that function a `try_new` instead, and remove the `assert!()` that might panic. I believe these values can be dynamically filled up by external actors of `editoast` so a more resilient code to `panic` might be preferable?

Reply:
The `PaginationStats` struct is meant to be used only in the responses and initialized after querying the results. The soundness of the pagination settings should already have been verified by then.

I don't understand what you mean by _"external actors"_. This struct can be created either by `trait PaginatedList` or by endpoints that wish to manually manage their pagination. 

Besides I'm not sure what error we'd like to expose to the issuer of the request as if one of this `assert!` fails, it's clearly a bug on our side.

Reply:
> The PaginationStats struct is meant to be used only in the responses and initialized after querying the results. The soundness of the pagination settings should already have been verified by then.

If it's already verified, then why `assert!()` here?

> I don't understand what you mean by "external actors". This struct can be created either by trait PaginatedList or by endpoints that wish to manually manage their pagination.

By external, I meant anything outside `editoast` (the `front` for example is an external client to `editoast`). 

> Besides I'm not sure what error we'd like to expose to the issuer of the request as if one of this assert! fails, it's clearly a bug on our side.

When you say "on our side", you mean the entire OSRD team or only the `editoast` team? In my mind, `editoast` itself should be protected against incorrect requests. So returning a `BAD REQUEST` in this case wouldn't feel wrong to me. Note that the 3 asserts might mean different kind of responses. The 2 first might mean `BAD REQUEST` if zero or negative. The third assert might be an error... or not, asking for a page that is further than the last page is semantically valid and might return an empty list.

Reply:
As discussed together, we won't be able to remove all the `assert!()` easily. The easiest solution we came up with was to use `NonZeroU64` instead of `u64` in some places, which should get rid of the 2 first `assert!()`.

Reply:
As discussed, since `NonZeroU64` is wrapper and not a drop-in replacement of `u64`, we prefer avoiding contaminating the whole stack with it. Too bad. I'm resolving this comment then.